Configuration from Web Browser

sebcaps opened the pull request:
This PR cancel and replace #71.
Even the purpose is the same, implementation is different and use a rest API architecture for better/easier extensibility. 
Other changes are : 
- new webserver library  (ESPAsyncWebServer vs esp library)
- html refactoring
fix #55 , fix-partially #63

sebcaps committed:
first commit

sebcaps committed:
first commit

sebcaps committed:
addMqttConfigHandling

sebcaps committed:
addMqttConfigHandling

sebcaps committed:
WebConfiguration

sebcaps committed:
WebConfiguration

sebcaps committed:
remove html file

sebcaps committed:
remove html file

sebcaps committed:
Delete AquaMQTT/data/test.html

sebcaps committed:
Merge branch 'OnlineConfig' of https://github.com/sebcaps/AquaMQTT into OnlineConfig

sebcaps committed:
finalversion

sebcaps committed:
finalversion

tspopp committed:
doc(readme): added discord for community and pcb trading possibility

tspopp committed:
doc(readme): fix discord link

tspopp commented on AquaMQTT/README.md:
````diff
@@ -34,37 +35,6 @@ Listener" and "Man-in-the-Middle."
-   constexpr EOperationMode OPERATION_MODE = EOperationMode::MITM;
-   ```
-
 ### Flashing
````
If we do not longer need configuration within the firmware, I will release pre-built images ready to be flashed e.g. also with esp web tool: https://esp.huhn.me/. Advances users still may want to build directly from source.

I will take care of the README and use your input as proposal :+1:

tspopp commented on AquaMQTT/data/index.html:
```diff
@@ -0,0 +1,191 @@
+                            <div class="input-group has-validation">
+                                <input class='form-control' id='ssid' type='text' name='ssid' required>
+                                <div class="invalid-feedback">
+                                    Please choose a username.
```
```suggestion
                                    Please enter a valid SSID.
```

tspopp commented on AquaMQTT/data/index.html:
```diff
@@ -0,0 +1,191 @@
+                    <h1>MQTT Configuration</h1>
+                    <form id='mqtt' method='POST' action='/api/mqtt'>
+                        <div class='form-group mb-3'>
+                            <label for='mqtt_server'>Server MQTT</label>
```
```suggestion
                            <label for='mqtt_server'>Server Address</label>
```

tspopp commented on AquaMQTT/data/index.html:
```diff
@@ -0,0 +1,191 @@
+                            <input class='form-control' id='mqtt_server' type='text' name='MQTT_Server'>
+                        </div>
+                        <div class='form-group mb-3'>
+                            <label for='mqtt_port'>MQTT server Port</label>
```
```suggestion
                            <label for='mqtt_port'>Server Port</label>
```

tspopp commented on AquaMQTT/data/index.html:
```diff
@@ -0,0 +1,191 @@
+                            <input class='form-control' id='mqtt_port' type='number' name='MQTT_ServerPort'>
+                        </div>
+                        <div class='form-group mb-3'>
+                            <label for='mqtt_user'>UserName</label>
```
```suggestion
                            <label for='mqtt_user'>User</label>
```

tspopp commented on AquaMQTT/data/web/css/bootstrap.min.css:
What is the purpose of these `.css` files?

tspopp commented on AquaMQTT/data/web/js/jquery-min.js:
Can you explain the background of choosing js/jquery?

sebcaps replied:
Actually with old version of bootstrap it was required, and it provide useful shortcuts and utility in other JS situation.
No more required by bootstrap. I mananed to handle JS needs without => I removed it ;-)

tspopp commented on AquaMQTT/include/config/Configuration.h:
I guess final goal of this PR would be removing this file entirely, right?

sebcaps replied:
Ultimately, yes. But don't took into account all parameters, defined, and maybe some does not worth UI configuration.
So I keep it.

tspopp commented on AquaMQTT/include/config/ExampleConfiguration.h:
Remove this file

tspopp commented on AquaMQTT/include/handler/Web.h:
```diff
@@ -0,0 +1,192 @@
+namespace web
+{
+
+const char HTTP_HEADER[] PROGMEM =
```
PROGMEM is deprecated, use constexpr.

```suggestion
constexpr char HTTP_HEADER[] =
```

tspopp commented on AquaMQTT/include/handler/Web.h:
```diff
@@ -0,0 +1,192 @@
+        "</ul></div>"
+        "</nav>";
+
+const char HTTP_WIFI[] PROGMEM =
```
constexpr

tspopp commented on AquaMQTT/include/handler/Web.h:
```diff
@@ -0,0 +1,192 @@
+        "name='save'>Save</button>"
+        "</div>";
+
+const char HTTP_MQTT[] PROGMEM =
```
constexpr

tspopp commented on AquaMQTT/include/handler/Web.h:
```diff
@@ -0,0 +1,192 @@
+        "name='save'>Save</button>"
+        "</div>";
+
+const char HTTP_AQUA[] PROGMEM =
```
constexpr

sebcaps replied:
Removed : no more necessary

tspopp commented on AquaMQTT/src/config/config.cpp:
```diff
@@ -0,0 +1,184 @@
+aquamqtt::MqttSettingStruct mqttSettings;
+aquamqtt::AquaMqttStruct    aquamqttSettings;
+
+IPAddress aquamqtt::parse_ip_address(const char* str)
```
Do you really need this utility function? Isn't there something provided from the framework?

sebcaps replied:
Removed : no more necessary

tspopp commented on AquaMQTT/src/config/config.cpp:
```diff
@@ -0,0 +1,184 @@
+        return false;
+    }
+    const char* ssid = doc["ssid"];
+    // double tmp = doc["tmp"];
+    // ConfigSettings.ssid = ssid;
+    // char pass_[30];
+    // // strcpy(pass_, doc["pass"]);
+    // ConfigSettings.password = doc["pass"];
+    // // char ip_[30];
+    // // strcpy(ip_, doc["pass");
+    // ConfigSettings.ipAddress = doc["ip"];
+    // // char mask_[30];
+    // // strcpy(mask_, doc["mask"]);
+    // ConfigSettings.ipMask = doc["mask"];
+    // // char gw_[30];
+    // // strcpy(gw_, doc["gw"]);
+    // ConfigSettings.ipGW = doc["gw"];
```
```suggestion
```

remove commented items

tspopp commented on AquaMQTT/src/main.cpp:
```diff
@@ -20,51 +24,88 @@ MQTTTask       mqttTask;
+        return;
+    }
+    // // Try to load config
+    if (!loadWifiConfig())
+    {
+        Serial.println("Wifi conf not loaded ....");
+    }
+    else
+    {
+        configOK = true;
+    }
+
+    if (configOK)
+    {
+        if (!wifiHandler.setup())
+        {
+            wifiHandler.setupAP();
+        }
+    }
+    else
+    {
+        Serial.println("Wifi set to Access Point Mode");
+        wifiHandler.setupAP();
+    }
```
```suggestion
    if (!loadWifiConfig())
    {
        Serial.println("Wifi conf not loaded ....");
         Serial.println("Wifi set to Access Point Mode");
        wifiHandler.setupAP();
    }
    else
    {
        if (!wifiHandler.setup())
        {
            wifiHandler.setupAP();
        }
    }

```

Do you really need the `configOK` field?

tspopp commented on AquaMQTT/src/main.cpp:
```diff
@@ -20,51 +24,88 @@ MQTTTask       mqttTask;
     .idle_core_mask = (1 << configNUM_CORES) - 1,
     .trigger_panic  = true,
 };
+bool                                   configOK = false;
```
```suggestion
```
see below

tspopp reviewed (changes requested):
Overall this looks very solid :+1:  I have not installed this yet, but will most probably try it out the next days. There are a few spelling things which we can resolve later. I did a first review of the code and I believe I like what I see and how you added this to the codebase.

I think we should prepare this PR to remove the entire Configuration from the firmware, so we will be able to provide firmware images which does not contain any custom configuration.

Still, I do not undestand the need for the javascript and the css files. If you could provide some quick explanation and background and maybe why you chosen that, would be very helpful.

Actually I thought about going along with https://github.com/tzapu/WiFiManager. What was the reason for not taking this library into accout?

sebcaps committed:
Fix PR comments

sebcaps committed:
Fix PR comments

sebcaps committed:
Update AquaMQTT/src/main.cpp

Co-authored-by: Thomas Popp <66408890+tspopp@users.noreply.github.com>

sebcaps committed:
Update AquaMQTT/src/main.cpp

Co-authored-by: Thomas Popp <66408890+tspopp@users.noreply.github.com>

sebcaps commented:
Hello, 
 All comments taken into account.
Actually, I did not know WifiManager and so recreate a similar and light implementation for wifi settings. (eg no static IP...)
From what I see it can handle more parameter than just wifi. 
Well let's say I somehow reinvented the wheel.

tspopp committed:
doc(readme): fix discord link

tspopp committed:
fix(next): fix option allow heat element and pv mode heat element (#75)

* fix(next): fix heating element enabled switch

* chore(version): bump version to v1.7.1

tspopp committed:
fix(next): fix parsing of operation type (#78)

mnederlof committed:
fix(aquamqtt): provide proper boiler brand to mqtt (#79)

tspopp committed:
fix(odyssee): fix setting operation type and mode

tspopp committed:
fix(next/odyssee): proper date overrides

tspopp committed:
test(protocol): unittest for hmi protocol overrides (next/legacy/odyssee)

tspopp committed:
chore(version): bump to v1.7.2

tspopp committed:
doc(devices): updated list of proofen devices

tspopp committed:
feat(protocol): parse and provide software version attributes (#81)

tspopp committed:
fix(serial): allow serial configuration from configuration file (#88)

Gwaboo committed:
doc(homeassistant): added guide to setup climate entity (#89)

* added a guide inside of HOMEASSISTANT.md` to setup a climat entity

* added custom climate card showcase images

* Updated HOMEASSISTANT.md with 3 showcase images from ./img

NickdeK committed:
fix(mitm): fix issue where invalid time overrides caused inoperative heatpumps

The mLastStateUpdate variable was declared as time_t (signed integer)
while being compared with millis() return values (unsigned long). This
caused incorrect behavior after millis() exceeded the maximum signed
integer value (2,147,483,647ms / ~24.8 days), causing time updates to
stop being sent.

Changed mLastStateUpdate to unsigned long to match the type returned by
millis() and ensure proper overflow handling in time comparisons.

tspopp committed:
chore(release): bump version to v1.8.0

tspopp committed:
doc(readme): incorporated various feedback, add thermor aeromax rt+

tspopp committed:
feat(climate): add climate component to home-assistant discovery (#100)

sebcaps committed:
Merge branch 'ApiStyle' of https://github.com/sebcaps/AquaMQTT into ApiStyle

phil-schneider commented:
I think this would be a real changer for this project.
People would not need to checkout git, compile themself and so one.
Currently every user needs to compile it themself.
With this: The user can upload a precompiled version. And then adjust the settings via a webinterface.
Simular to WLED.

sebcaps commented:
New version with all the parameters in progress 
<img width="1064" height="791" alt="image" src="https://github.com/user-attachments/assets/36417db3-4701-4341-9b2e-94e184584d5d" />
<img width="1139" height="694" alt="image" src="https://github.com/user-attachments/assets/f3e38beb-8096-4520-b2ed-b38d94a70cb1" />
<img width="1193" height="356" alt="image" src="https://github.com/user-attachments/assets/2a6abf10-18cd-4ed7-ae74-372ab136196a" />
<img width="945" height="417" alt="image" src="https://github.com/user-attachments/assets/a210fcc4-c975-44a9-bacf-9d397b054a5b" />

Almost good from a web / api perspective. 
Will need more testing from esp32 perspective. Not a precise idea on how-to without connecting to a real device. 
In addition I think it would be good to clamp some parameter like Kalman filter params, update intervalls; but no clue on limits.

Any hints about these two points ?

tspopp commented:
> Will need more testing from esp32 perspective. Not a precise idea on how-to without connecting to a real device.

Have not looked yet into the code, but if you can move all the business logic unrelated to the Arduino framework into a separate internal library, you can use the `native` platformio environment to write and run unittests. I did the same for a few protocol test I needed to write back then. Maybe this helps?

> In addition I think it would be good to clamp some parameter like Kalman filter params, update intervalls; but no clue on limits.

Maybe me make these kalman parameters just const, not changeable through the web-if. Only very advanced users will change them anyway. What do you think?

sebcaps commented:
> > Will need more testing from esp32 perspective. Not a precise idea on how-to without connecting to a real device.
> 
> Have not looked yet into the code, but if you can move all the business logic unrelated to the Arduino framework into a separate internal library, you can use the `native` platformio environment to write and run unittests. I did the same for a few protocol test I needed to write back then. Maybe this helps?

Will have a look to unittests, hope I'll manage to handle them...

> > In addition I think it would be good to clamp some parameter like Kalman filter params, update intervalls; but no clue on limits.
> 
> Maybe me make these kalman parameters just const, not changeable through the web-if. Only very advanced users will change them anyway. What do you think?

Actually, I added all parameters from Configuration.h (except WATCHDOG_TIMEOUT_MS, WIFI_RECONNECT_CYCLE_S, MQTT_MAX_TOPIC_SIZE, MQTT_MAX_PAYLOAD_SIZE which are unlikely to change) to  almost remove the entire Configuration from the firmware.
I will re-add MQTT_FILTER_TEMPERATURE_NOISE, KALMAN_MEA_E, KALMAN_EST_E and KALMAN_Q.

Do you envision any other settings that are not very relevant from web config ?
